pythonforandroid/recipes/python3/__init__.py

```diff
@@ -33,13 +33,53 @@
 
 
 class Python3Recipe(TargetPythonRecipe):
+    '''
+    .. note::
+        In order to build certain python modules, we need to add some extra
+        recipes to our build requirements:
+
+            - ctypes: you must add the recipe for ``libffi``.
+    '''
     version = '3.7.1'
     url = 'https://www.python.org/ftp/python/{version}/Python-{version}.tgz'
     name = 'python3'
 
     depends = ['hostpython3']
     conflicts = ['python3crystax', 'python2']
-    # opt_depends = ['openssl', 'sqlite3']
+    opt_depends = ['libffi']  # 'openssl', 'sqlite3'
+
+    configure_args = (
+        '--host={android_host}',
+        '--build={android_build}',
+        '--enable-shared',
+        '--disable-ipv6',
+        'ac_cv_file__dev_ptmx=yes',
+        'ac_cv_file__dev_ptc=no',
+        '--without-ensurepip',
+        'ac_cv_little_endian_double=yes',
+        '--prefix={prefix}',
+        '--exec-prefix={exec_prefix}')
+
+    def set_libs_flags(self, env, arch):
+        '''Takes care to properly link libraries with python depending on our
+        requirements and the attribute :attr:`opt_depends`.
+        '''
+        def add_flags(include_flags, link_flags):
+            env['CPPFLAGS'] = env.get('CPPFLAGS', '') + include_flags
+            env['LDFLAGS'] = env.get('LDFLAGS', '') + link_flags
+
+        if 'libffi' in self.ctx.recipe_build_order:
+            info('Activating flags for libffi')
+            recipe = Recipe.get_recipe('libffi', self.ctx)
+            include = ' -I' + ' -I'.join(recipe.get_include_dirs(arch))
+            ldflag = ' -L' + join(recipe.get_build_dir(arch.arch),
+                                  recipe.get_host(arch), '.libs') + ' -lffi'
+            add_flags(include, ldflag)
+            # libffi needs some extra configurations
+            env['LIBFFI_CFLAGS'] = env.get('CFLAGS', '') + include
+            env['LIBFFI_LIBS'] = ldflag
+            self.configure_args += ('--with-system-ffi',)
```

Comment:
I think `--with-system-ffi` should be unnecessary on linux, and on osx it seems cpython adds `/usr/include/ffi` to the include path automatically. Have you tried on osx, and with omitting this argument, I'm not sure if I've interpreted it correctly?

It doesn't really matter too much, just wanting to understand what's going on.

Reply:
Again, You are right, so I will revert all the changes affecting the attribute configure_args, there is no need to touch it. So many thanks for your clinical eyes @inclement with all the changes this becomes a lot more cleaner  :blush: